### PR TITLE
Introduce IMEX Runge-Kutta time steppers into existing integrators module 

### PR DESCRIPTION
**Pull request by mikaem**
## Summary

Adds diagonally-implicit IMEX Runge-Kutta (IMEX-RK) time integration to `jaxfun.integrators`, generalizing the existing ad hoc IMEX-Euler scheme in `BackwardEuler` to a generic, Butcher-tableau-driven family. Motivated by wanting higher-order (>1) implicit-explicit schemes for stiff-linear / nonstiff-nonlinear semi-discrete PDE systems, without changing the framework's existing scalar-field scope (coupled multi-field systems are intentionally out of scope for this PR — see Additional Notes).

## Types of Changes

- [x] New feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactor / cleanup
- [ ] Build / CI
- [ ] Other (describe):

## Related Issues

Fixes # \
Refs #

## Description of Changes

- **`src/jaxfun/integrators/tableau.py`** (new): `ButcherTableau` and `IMEXTableau` dataclasses with structural validation (square `A`, row-sum-consistency `sum(A[i]) == c[i]`, strictly-lower-triangular explicit table, lower-triangular/diagonally-implicit implicit table). `IMEXTableau` exposes `stages`, `distinct_diagonal_coeffs` (drives how many implicit system matrices need caching), and stiff-accuracy classification: `explicit_is_stiffly_accurate`, `implicit_is_stiffly_accurate`, and the combined `is_stiffly_accurate` (both).

  Seven named schemes, each transcribed from a primary/authoritative source and independently verified (row-sum consistency, triangularity, and — where feasible — exact `fractions.Fraction` arithmetic) before being trusted, not just copied:
  - `IMEX_EULER` — 2-stage, order 1; reduces exactly to `BackwardEuler` (tested).
  - `IMEX_SSP2_222` — Pareschi & Russo IMEX-SSP2(2,2,2), order 2, L-stable.
  - `ARK2_1_3L2SA`, `ARK3_2_4L2SA`, `ARK4_3_6L2SA`, `ARK5_4_8L2SA` — Kennedy-Carpenter ARK family, orders 2/3/4/5, L-stable, implicit-part-only stiffly accurate. Coefficients transcribed from SUNDIALS ARKODE's raw source, not secondary summaries (an earlier AI-summarized fetch subtly misread PETSc's `TSARKIMEXARS443` default-weight convention — see `ARS443`'s docstring — which is why the later schemes were pulled from raw `.def` source and cross-checked with exact rational arithmetic instead).
  - `ARS222`, `ARS443` — Ascher-Ruuth-Spiteri (1997), orders 2/3, L-stable, and *globally* stiffly accurate (both tableaux independently satisfy `A[-1] == b`), enabling the full last-stage-reuse shortcut.

- **`src/jaxfun/integrators/imex_rk.py`** (new): `IMEXRungeKutta(BaseIntegrator)`. `setup(dt)` caches one implicit system operator (`mass_operator - dt*a_ii*linear_operator`) per distinct diagonal Butcher coefficient. `step()` runs a Python-level (trace-time-unrolled) stage loop reusing `BackwardEuler`'s per-stage solve pattern and `ETDRK4`'s per-stage nonlinear-evaluation pattern, then picks one of three final-combination paths based on the tableau's static stiff-accuracy classification:
  1. **Globally stiffly accurate**: return the last stage directly.
  2. **Implicit-only stiffly accurate**: fold the `b_i`-weighted combination algebraically into the last stage (derived from `implicit.A[-1] == b_i`), leaving only a `(b_e - explicit.A[-1])`-weighted correction over already-cached nonlinear stage values — no linear-operator terms needed at all.
  3. **General**: the full two-sided weighted combination.

  Each path also skips computing the last stage's nonlinear/linear evaluations when the chosen path doesn't need them, avoiding wasted `linear_operator @` matvecs (and, in the fully-GSA case, a wasted nonlinear evaluation) that the naive single-path implementation would otherwise always compute and discard.

- **`src/jaxfun/integrators/__init__.py`**: exports `IMEXRungeKutta`, `ButcherTableau`, `IMEXTableau`, and the seven named schemes.

- **`examples/ginzburg_landau_imex.py`** (new): 2D complex-valued Ginzburg-Landau equation on a Fourier tensor-product space, integrated with `IMEXRungeKutta(tableau=ARS443)` — a real nonlinear/stiff usage example beyond the unit tests.

- Two small, unrelated drive-by fixes bundled in the same commit: `examples/cahn_hilliard2D_etdrk4.py` updated to the current `TensorProduct` API (was using the older `TensorProductSpace` name), and `DiaMatrix.__repr__` now uses `self.__class__.__name__` instead of a hardcoded `"DiaMatrix"`.

Design note on scope: mixing an arbitrary explicit tableau with an arbitrary implicit tableau does **not** generally preserve order above 1 — orders ≥2 require coupling/cross conditions between the two tables (e.g. `Σ bᵢᴱcᵢᴵ = 1/2`), so every scheme here is a specific, jointly-designed, published pair, not an ad hoc combination.

## Screenshots / Logs (if applicable)

```
$ uv run pytest tests/ -q
840 passed, 8 skipped
$ uv run ruff check src/jaxfun/integrators/ tests/integrators/
All checks passed!
$ uv run ty check src/jaxfun/integrators/
All checks passed!
```

## Breaking Changes

- [ ] Yes (describe below)
- [x] No

Purely additive: no existing files in `src/jaxfun/integrators/` (`base.py`, `backward_euler.py`, `etdrk4.py`, `rk4.py`) were modified, and all pre-existing tests pass unchanged.

## Checklist

- [x] Tests added or updated
- [ ] Documentation updated (README, examples, docstrings) — docstrings/example added; no README changes
- [x] Added type hints
- [x] Linting & formatting pass locally (`pre-commit run --all-files`)
- [x] All new and existing tests pass (`uv run pytest`)
- [ ] Coverage does not decrease — not explicitly measured, but new modules are fully covered by new tests
- [x] Git history is clean (squash/fixup before merge)

## Additional Notes

- Coupled/multi-field time-dependent systems (e.g. a genuine two-field Cahn-Hilliard split, or Navier-Stokes-style coupling) were explored but deliberately **not** included in this PR. An initial design using `CartesianProductSpace`/`BlockMatrix`/`BlockArray` was considered and rejected as too heavy; the preferred direction is a "segregated" approach (each field keeps its own scalar integrator; cross-field terms enter as known/lagged data in another field's nonlinear evaluator). That's follow-up work, not blocking this PR, since `IMEXRungeKutta` is useful standalone for scalar equations today.
- New tests (`tests/integrators/test_tableau.py`, `tests/integrators/test_imex_rk.py`) include: structural/consistency checks for every tableau; `IMEX_EULER` reducing exactly to `BackwardEuler`; empirical convergence-order checks (halving `dt` on a purely-linear problem) for `ARK2_1_3L2SA` (~2.0), `ARS222` (~2.0), `ARK3_2_4L2SA` (~2.9), `ARS443` (~2.9), `ARK4_3_6L2SA` (~3.9), and `ARK5_4_8L2SA` (~5.0); an `ARK4_3_6L2SA` vs. `ETDRK4` cross-check on a KdV soliton; and an `IMEX_SSP2_222` accuracy smoke test (its DIRK companion is only 1st-order standalone, so a pure-linear order test isn't meaningful for it — see the test docstring).

### CODE REVIEW

**Comment by augustfe** [comment]
Questions before going through the changes:
- Is adaptive timestepping implemented? If not, it may be simple to use `Optimistix`, in much the same way `Diffrax` works, instead of implementing everything from scratch.
- The butcher tableau method should probably be extended to cover ETDRK, as I think only ETDRK4 is currently covered?

**Comment by mikaem** [comment]
> Questions before going through the changes:
> 
> * Is adaptive timestepping implemented? If not, it may be simple to use `Optimistix`, in much the same way `Diffrax` works, instead of implementing everything from scratch.

No. Adaptive not implemented yet. Still much room for improvements:-)

> * The butcher tableau method should probably be extended to cover ETDRK, as I think only ETDRK4 is currently covered?

ETDRK/ETDRK4 are special methods, only intended for Fourier, right? They don't fit in the Butcher tableau approach, where everything falls out from the tableau numbers.

**Comment by augustfe** [comment]
> > Questions before going through the changes:
> 
> > 
> 
> > * Is adaptive timestepping implemented? If not, it may be simple to use `Optimistix`, in much the same way `Diffrax` works, instead of implementing everything from scratch.
> 
> 
> 
> No. Adaptive not implemented yet. Still much room for improvements:-)
> 
> 
> 
> > * The butcher tableau method should probably be extended to cover ETDRK, as I think only ETDRK4 is currently covered?
> 
> 
> 
> ETDRK/ETDRK4 are special methods, only intended for Fourier, right? They don't fit in the Butcher tableau approach, where everything falls out from the tableau numbers.
> 
> 

ETDRK is essentially just RK with some fancy pre-processing, I think the step part just follows RK4. I have seen ETDRK2/3 previously

**Comment by mikaem** [comment]
> > > Questions before going through the changes:
> > 
> > 
> > > 
> > 
> > 
> > > * Is adaptive timestepping implemented? If not, it may be simple to use `Optimistix`, in much the same way `Diffrax` works, instead of implementing everything from scratch.
> > 
> > 
> > No. Adaptive not implemented yet. Still much room for improvements:-)
> > > * The butcher tableau method should probably be extended to cover ETDRK, as I think only ETDRK4 is currently covered?
> > 
> > 
> > ETDRK/ETDRK4 are special methods, only intended for Fourier, right? They don't fit in the Butcher tableau approach, where everything falls out from the tableau numbers.
> 
> ETDRK is essentially just RK with some fancy pre-processing, I think the step part just follows RK4. I have seen ETDRK2/3 previously

Ok. I see from https://arxiv.org/pdf/1604.08900 that it fits in a Butcher tableau approach and it should probably be reimplemented using a common baseclass for ETDRK2/3/4. But it is still very different from the imex solvers and should probably not touch the IMEXRungeKutta baseclass used in this PR, right?

File: src/jaxfun/integrators/tableau.py
Comment (augustfe) [inline src/jaxfun/integrators/tableau.py:1]: Why are the tableaus specified as tuples? Wouldn't it make more sense to have them as Arrays? Surely it would make some of the checking/indexing simpler

**Comment by augustfe** [comment]
> > > > Questions before going through the changes:
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > * Is adaptive timestepping implemented? If not, it may be simple to use `Optimistix`, in much the same way `Diffrax` works, instead of implementing everything from scratch.
> 
> > > 
> 
> > > 
> 
> > > No. Adaptive not implemented yet. Still much room for improvements:-)
> 
> > > > * The butcher tableau method should probably be extended to cover ETDRK, as I think only ETDRK4 is currently covered?
> 
> > > 
> 
> > > 
> 
> > > ETDRK/ETDRK4 are special methods, only intended for Fourier, right? They don't fit in the Butcher tableau approach, where everything falls out from the tableau numbers.
> 
> > 
> 
> > ETDRK is essentially just RK with some fancy pre-processing, I think the step part just follows RK4. I have seen ETDRK2/3 previously
> 
> 
> 
> Ok. I see from https://arxiv.org/pdf/1604.08900 that it fits in a Butcher tableau approach and it should probably be reimplemented using a common baseclass for ETDRK2/3/4. But it is still very different from the imex solvers and should probably not touch the IMEXRungeKutta baseclass used in this PR, right?

Yes that sounds right

File: src/jaxfun/integrators/tableau.py
Comment (mikaem) [reply src/jaxfun/integrators/tableau.py:1]: Several reasons. 1) tableau.py does not import jax. float in tableau.py are float64 regardless of jax settings 2) The IMEXTableau are static, compile-time constants not runtime data, good for jit 3) Arrays are unhashable

**Comment by mikaem** [comment]
Merge @augustfe ?

**Comment by augustfe** [comment]
> Merge @augustfe ?

Yes, sounds good
